8339475: Clean up return code handling for pthread calls in library coding

MBaesken opened the pull request:
It has been discussed that checking the return value of pthread_create should be done.
See the discussion here https://github.com/openjdk/jdk/pull/20812 about the splashscreen coding .

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8339475](https://bugs.openjdk.org/browse/JDK-8339475): Clean up return code handling for pthread calls in library coding (**Bug** - P4)


### Reviewers
 * [Christoph Langer](https://openjdk.org/census#clanger) (@RealCLanger - **Reviewer**)
 * [Julian Waters](https://openjdk.org/census#jwaters) (@TheShermanTanker - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21182/head:pull/21182` \
`$ git checkout pull/21182`

Update a local copy of the PR: \
`$ git checkout pull/21182` \
`$ git pull https://git.openjdk.org/jdk.git pull/21182/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21182`

View PR using the GUI difftool: \
`$ git pr show -t 21182`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21182.diff">https://git.openjdk.org/jdk/pull/21182.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21182#issuecomment-2373864027)

MBaesken commented:
I found some strange handling of the return value on macOS.
Here we check already the return value of pthread_create ; but then it seems we add the errno twice when calling JLI_ReportErrorMessageSys, once in the calling params and once in JLI_ReportErrorMessageSys itself.
Should we better call  with the return value of pthread_create in the error case? Or just with a string without additional parameters?
The comment about Solaris 8 is also very outdated, don't think we need that any more in jdk24.

```
src/java.base/macosx/native/libjli/java_md_macosx.m 

312     if (pthread_create(&main_thr, NULL, &apple_main, &args) != 0) {
313         JLI_ReportErrorMessageSys("Could not create main thread: %s\n", strerror(errno));

src/java.base/unix/native/libjli/java_md_common.c

244  JLI_ReportErrorMessageSys(const char* fmt, ...) {
245      va_list vl;
246      char *emsg;
247  
248      /*
249       * TODO: its safer to use strerror_r but is not available on
250       * Solaris 8. Until then....
251       */
252      emsg = strerror(errno);
253      if (emsg != NULL) {
254          fprintf(stderr, "%s\n", emsg);
255      }
```

RealCLanger commented on src/java.desktop/macosx/native/libsplashscreen/splashscreen_sys.m:
```diff
@@ -276,6 +276,9 @@ static int isInAquaSession() {
     int rslt = pthread_attr_init(&attr);
     if (rslt != 0) return;
     rc = pthread_create(&thr, &attr, SplashScreenThread, (void *) splash);
```
For the sake of symmetry, could you rename rc to rslt here?

MBaesken replied:
Sure, I changed it .

RealCLanger reviewed (changes requested):
Looks good overall, just one nit.

I guess it's ok to print a warning to stderr in this case.

RealCLanger commented:
> JLI_ReportErrorMessageSys



> I found some strange handling of the return value on macOS. Here we check already the return value of pthread_create ; but then it seems we add the errno twice when calling JLI_ReportErrorMessageSys, once in the calling params and once in JLI_ReportErrorMessageSys itself. Should we better call with the return value of pthread_create in the error case? Or just with a string without additional parameters? The comment about Solaris 8 is also very outdated, don't think we need that any more in jdk24.
> 
> ```
> src/java.base/macosx/native/libjli/java_md_macosx.m 
> 
> 312     if (pthread_create(&main_thr, NULL, &apple_main, &args) != 0) {
> 313         JLI_ReportErrorMessageSys("Could not create main thread: %s\n", strerror(errno));
> 
> src/java.base/unix/native/libjli/java_md_common.c
> 
> 244  JLI_ReportErrorMessageSys(const char* fmt, ...) {
> 245      va_list vl;
> 246      char *emsg;
> 247  
> 248      /*
> 249       * TODO: its safer to use strerror_r but is not available on
> 250       * Solaris 8. Until then....
> 251       */
> 252      emsg = strerror(errno);
> 253      if (emsg != NULL) {
> 254          fprintf(stderr, "%s\n", emsg);
> 255      }
> ```

Yes, I think the call to java_md_macosx.m seems redundant. And I would remove the comment in java_md_common.c. strerror is used at various places in JDK coding and there's actually only one place that uses strerror_r, in jni_util_md.c

MBaesken commented:
>Yes, I think the call to java_md_macosx.m seems redundant. And I would remove the comment in java_md_common.c. strerror is >used at various places in JDK coding and there's actually only one place that uses strerror_r, in jni_util_md.c

Thanks for the comment;  should I change it in this PR or separately ?

TheShermanTanker commented:
> I found some strange handling of the return value on macOS. Here we check already the return value of pthread_create ; but then it seems we add the errno twice when calling JLI_ReportErrorMessageSys, once in the calling params and once in JLI_ReportErrorMessageSys itself. Should we better call with the return value of pthread_create in the error case? Or just with a string without additional parameters? The comment about Solaris 8 is also very outdated, don't think we need that any more in jdk24.
> 
> ```
> src/java.base/macosx/native/libjli/java_md_macosx.m 
> 
> 312     if (pthread_create(&main_thr, NULL, &apple_main, &args) != 0) {
> 313         JLI_ReportErrorMessageSys("Could not create main thread: %s\n", strerror(errno));
> 
> src/java.base/unix/native/libjli/java_md_common.c
> 
> 244  JLI_ReportErrorMessageSys(const char* fmt, ...) {
> 245      va_list vl;
> 246      char *emsg;
> 247  
> 248      /*
> 249       * TODO: its safer to use strerror_r but is not available on
> 250       * Solaris 8. Until then....
> 251       */
> 252      emsg = strerror(errno);
> 253      if (emsg != NULL) {
> 254          fprintf(stderr, "%s\n", emsg);
> 255      }
> ```

Ah, that problem with JLI_ReportErrorMessageSys. I recall that me and Thomas reported a possibly related issue for it ages ago
https://bugs.openjdk.org/browse/JDK-8292016

MBaesken commented:
Interesting, the errno is not even set on macOS according to the man page, at least for pthread_detach
https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/pthread_detach.3.html
"Note that the function does not change the value of errno, as it did for some drafts of the standard. "

MBaesken commented:
I removed the very old Solaris 8 comment.
I adjusted the macOS related code mentioned above.
For pthread_detach, I switched to  JLI_ReportErrorMessage , because the manpage says that errno is not set; so better do not print it in the failure case on macOS.

MBaesken commented:
Maybe "Check return value of pthread_create in jdk coding"  does not fit any more for the bug/PR name now ? Any good suggestions?

RealCLanger reviewed (approved):
LGTM.

The bug title could be changed into something like "Clean up return code handling for pthread calls in library coding"

MBaesken commented:
Thanks for the review !  I changed the title too .

Julian, are you fine with the latest rev. ?

TheShermanTanker reviewed (approved):
I am ok with this. I wonder what @tstuefe and @dholmes-ora think about the JLI\_ReportErrorMessageSys changes though. I recall one of them preferring strerror\_r over strerror

MBaesken commented:
> I am ok with this. I wonder what @tstuefe and @dholmes-ora think about the JLI_ReportErrorMessageSys changes though. I recall one of them preferring strerror_r over strerror

Hi Julian, in HS we use a large array of error strings :
8148425: strerror() function is not thread-safe
https://hg.openjdk.org/jdk9/jdk9/hotspot/rev/1537c752a7f5

In case we want to replace strerror in java_md_common.c (by strerror_r or by an array like we use in HS), that would be another separate issue.

MBaesken commented:
/integrate

MBaesken commented on src/java.base/macosx/native/libjli/java_md_macosx.m:
```diff
@@ -309,12 +310,14 @@ static void MacOSXStartup(int argc, char *argv[]) {
-        JLI_ReportErrorMessageSys("Could not create main thread: %s\n", strerror(errno));
+    rc = pthread_create(&main_thr, NULL, &apple_main, &args);
+    if (rc != 0) {
+        JLI_ReportErrorMessageSys("Could not create main thread, return code: %s\n", rc);
```
Thanks, I'll look into this next week.
